runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/StreamPoolHeartbeatSenderTest.java

```diff
@@ -0,0 +1,113 @@
+    heartbeatSender.sendHeartbeats(heartbeatsBuilder.build());
+    assertEquals(1, server.getGetDataRequests().size());
+  }
+
+  @Test
+  public void sendsHeartbeatsOnDedicatedStream() {
+    FakeWindmillServer dedicatedServer =
+        new FakeWindmillServer(new ErrorCollector(), c -> Optional.empty());
+    FakeWindmillServer getDataServer =
+        new FakeWindmillServer(new ErrorCollector(), c -> Optional.empty());
+
+    FixedGlobalConfigHandle configHandle =
+        new FixedGlobalConfigHandle(
+            StreamingGlobalConfig.builder()
+                .setUserWorkerJobSettings(
+                    UserWorkerRunnerV1Settings.newBuilder()
+                        .setUseSeparateWindmillHeartbeatStreams(true)
+                        .build())
+                .build());
+    StreamPoolHeartbeatSender heartbeatSender =
+        StreamPoolHeartbeatSender.Create(
+            WindmillStreamPool.create(
+                1, Duration.standardSeconds(10), dedicatedServer::getDataStream),
+            WindmillStreamPool.create(
+                1, Duration.standardSeconds(10), getDataServer::getDataStream),
+            configHandle);
+    Heartbeats.Builder heartbeatsBuilder = Heartbeats.builder();
+    heartbeatsBuilder
+        .heartbeatRequestsBuilder()
+        .put("key", HeartbeatRequest.newBuilder().setWorkToken(123).build());
+    heartbeatSender.sendHeartbeats(heartbeatsBuilder.build());
+    assertEquals(1, dedicatedServer.getGetDataRequests().size());
+    assertEquals(0, getDataServer.getGetDataRequests().size());
+  }
+
+  @Test
+  public void sendsHeartbeatsOnGetDataStream() {
+    FakeWindmillServer dedicatedServer =
+        new FakeWindmillServer(new ErrorCollector(), c -> Optional.empty());
+    FakeWindmillServer getDataServer =
+        new FakeWindmillServer(new ErrorCollector(), c -> Optional.empty());
+
+    FixedGlobalConfigHandle configHandle =
+        new FixedGlobalConfigHandle(
+            StreamingGlobalConfig.builder()
+                .setUserWorkerJobSettings(
+                    UserWorkerRunnerV1Settings.newBuilder()
+                        .setUseSeparateWindmillHeartbeatStreams(false)
+                        .build())
+                .build());
+    StreamPoolHeartbeatSender heartbeatSender =
+        StreamPoolHeartbeatSender.Create(
+            WindmillStreamPool.create(
+                1, Duration.standardSeconds(10), dedicatedServer::getDataStream),
+            WindmillStreamPool.create(
+                1, Duration.standardSeconds(10), getDataServer::getDataStream),
+            configHandle);
+    Heartbeats.Builder heartbeatsBuilder = Heartbeats.builder();
+    heartbeatsBuilder
+        .heartbeatRequestsBuilder()
+        .put("key", HeartbeatRequest.newBuilder().setWorkToken(123).build());
+    heartbeatSender.sendHeartbeats(heartbeatsBuilder.build());
+    assertEquals(0, dedicatedServer.getGetDataRequests().size());
+    assertEquals(1, getDataServer.getGetDataRequests().size());
```

Comment:
then set the config to different setting and verify it goes to the other server for next heartbeat sent on the same heartbeatSender object?

Reply:
Done.